watchdog/src/api_access.rs

```diff
@@ -0,0 +1,102 @@
+use crate::health::{HealthStatus, HeightStatus};
+use crate::print;
+use candid::CandidType;
+use ic_btc_interface::{Config as BitcoinCanisterConfig, Flag, SetConfigRequest};
+
+/// Captures the expected and the actual value of the Bitcoin canister API access flag.
+/// They might be different if the Bitcoin canister is not reachable.
+#[derive(Clone, Debug, CandidType)]
+pub struct ApiAccess {
+    /// Expected value of the Bitcoin canister API access flag.
+    pub target: Option<Flag>,
+
+    /// Actual value of the Bitcoin canister API access flag.
+    pub actual: Option<Flag>,
+}
+
+impl ApiAccess {
+    pub fn new() -> Self {
+        Self {
+            target: None,
+            actual: None,
+        }
+    }
+
+    /// Checks if the target and actual API access flags are in sync.
+    pub fn is_in_sync(&self) -> bool {
+        self.target == self.actual
+    }
+}
+
+impl Default for ApiAccess {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Calculates the target value of the Bitcoin canister API access flag.
+fn calculate_target(health: HealthStatus) -> Option<Flag> {
+    match health.height_status {
+        HeightStatus::Ok => Some(Flag::Enabled),
+        HeightStatus::Behind | HeightStatus::Ahead => Some(Flag::Disabled),
+        HeightStatus::NotEnoughData => None,
+    }
+}
+
+/// Fetches the Bitcoin canister config.
+async fn get_bitcoin_canister_config() -> Option<BitcoinCanisterConfig> {
+    let id = crate::storage::get_config().bitcoin_canister_principal;
+    let result = ic_cdk::api::call::call(id, "get_config", ()).await;
+    result
+        .map(|(config,)| config)
+        .map_err(|err| print(&format!("Error getting Bitcoin canister config: {:?}", err)))
+        .ok()
+}
+
+/// Fetches the actual API access flag and calculates the target value.
+async fn fetch_actual_and_calculate_target_api_access() {
+    let target = calculate_target(crate::health::health_status());
+
+    let bitcoin_canister_config = get_bitcoin_canister_config().await;
+    let actual = bitcoin_canister_config.map(|config| config.api_access);
+    if actual.is_none() {
+        print("Error getting Bitcoin canister config: api_access is None");
+    }
+
+    crate::storage::set_api_access(ApiAccess { target, actual });
+}
+
+/// Updates the API access flag in the Bitcoin canister.
+async fn update_api_access(target: Option<Flag>) {
+    let id = crate::storage::get_config().bitcoin_canister_principal;
+    let set_config_request = SetConfigRequest {
+        api_access: target,
+        ..Default::default()
+    };
+    let result = ic_cdk::api::call::call(id, "set_config", (set_config_request,)).await;
+    match result {
+        Ok(()) => (),
+        Err(err) => {
+            print(&format!("Error setting Bitcoin canister config: {:?}", err));
+        }
+    }
+}
+
+/// Synchronizes the API access flag of the Bitcoin canister, attempting a limited number of times.
+pub async fn synchronise_api_access() {
```

Comment:
The one point I'm not entirely sure about is whether we want to immediately turn on the flag again if we believe that the canister is again in sync with explorers.

I think disabling the access is totally fine (that's the goal of this work) but enabling automatically without someone checking why we had to disable in the first place and whether there is a deeper issue with the bitcoin canister is not so clear imo. @ielashi wdyt?

Reply:
Do you see a particular danger in turning it on again should we be synced again? As I think about, I don't see how that can be dangerous.

There's a bit of an annoyance with the current approach though, and that is that the watchdog is a bit too persistent in updating the config. For example, what if we wanted to manually disable the API even though we're synced because we discovered something else being broken? In that case, the watchdog canister would immediately enable the API again.

Maybe the watchdog should be more gentle, and only attempt to update the config once should the `target` change, and not any other time.

Reply:
Another idea: add a latch to the bitcoin canister that on every manual API turn off would automatically drop watchdog permissions for changing the API access.

So the maintenance steps would be:
- manually disable bitcoin canister API (automatically removes watchdog permissions)
- update bitcoin canister
- manually enable bitcoin canister API access
- (manually or automatically with the previous step) grant permissions for watchdog

Reply:
> Alternatively we can add a turn on/off button for a watchdog canister. So for every bitcoin canister it would require to run the watchdog off, do the thing on bitcoin canister and then turn the watchdog canister on again.

Right, that 's actually already possible with our current configuration. Whenever we want to manually override the watchdog, we can also set the `watchdog_canister` in the bitcoin canister's configuration to be `none`, and then it won't listen to the watchdog commands, and we can reset it afterwards. I'll add that to the instructions in the runbook.

Reply:
> Do you see a particular danger in turning it on again should we be synced again? As I think about, I don't see how that can be dangerous.

Not necessarily dangerous but it could also lead to some turn off/turn on loop which would be also annoying. I think it's mostly about making sure that we look into cases where the bitcoin canister fell behind and understand what happened as opposed to just ignore because after a few minutes it was again syncing and then we re-enabled access. I suppose if we get notified whenever the watchdog needs to intervene that would also be ok, we can look into issues/improvements afterwards.

> For example, what if we wanted to manually disable the API even though we're synced because we discovered something else being broken? In that case, the watchdog canister would immediately enable the API again.

Yes, that's another good point. As discussed, we can, in that case, remove the permission for `watchdog_canister` and manually disable the apis.

Reply:
To conclude the discussion, I suppose, we agreed that the bitcoin canister should handle watchdog's permissions, either manually or automatically. Since there is no action points for the watchdog I'm marking this conversation resolved. Feel free to reopen.